### PR TITLE
Fix null error on live photo import

### PR DESCRIPTION
**Pull request by icager**
Hi - sorry in advance if I haven't done enough on this PR. I've tried to add a test to demonstrate this bugfix but not being particularly familiar with PHP running it locally was beyond me, so I'm hoping an automated build will catch it if it doesn't work.

The bug I've attempted to fix is an API error when uploading the second part of a live photo pair (either photo or video). Screenshot attached, and stacktrace as follows:
```
ErrorException:
Attempt to read property &quot;type&quot; on null

at /tmp/Lychee/app/Actions/Photo/Strategies/StrategyPhotoBase.php:84
    at Illuminate\Foundation\Bootstrap\HandleExceptions-&gt;handleError()
(/tmp/Lychee/app/Actions/Photo/Strategies/StrategyPhotoBase.php:84)
    at App\Actions\Photo\Strategies\StrategyPhotoBase-&gt;findLivePartner()
(/tmp/Lychee/app/Actions/Photo/Create.php:119)
    at App\Actions\Photo\Create-&gt;add()
(/tmp/Lychee/app/Http/Controllers/PhotoController.php:105)
    at App\Http\Controllers\PhotoController-&gt;add()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Controller.php:54)
    at Illuminate\Routing\Controller-&gt;callAction()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:45)
    at Illuminate\Routing\ControllerDispatcher-&gt;dispatch()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Route.php:254)
    at Illuminate\Routing\Route-&gt;runController()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Route.php:197)
    at Illuminate\Routing\Route-&gt;run()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:695)
    at Illuminate\Routing\Router-&gt;Illuminate\Routing\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/app/Http/Middleware/UploadCheck.php:45)
    at App\Http\Middleware\UploadCheck-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php:50)
    at Illuminate\Routing\Middleware\SubstituteBindings-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php:78)
    at Illuminate\Foundation\Http\Middleware\VerifyCsrfToken-&gt;handle()
(/tmp/Lychee/app/Http/Middleware/VerifyCsrfToken.php:61)
    at App\Http\Middleware\VerifyCsrfToken-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Session/Middleware/AuthenticateSession.php:39)
    at Illuminate\Session\Middleware\AuthenticateSession-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php:49)
    at Illuminate\View\Middleware\ShareErrorsFromSession-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:121)
    at Illuminate\Session\Middleware\StartSession-&gt;handleStatefulRequest()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:64)
    at Illuminate\Session\Middleware\StartSession-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php:37)
    at Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php:67)
    at Illuminate\Cookie\Middleware\EncryptCookies-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103)
    at Illuminate\Pipeline\Pipeline-&gt;then()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:697)
    at Illuminate\Routing\Router-&gt;runRouteWithinStack()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:672)
    at Illuminate\Routing\Router-&gt;runRoute()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:636)
    at Illuminate\Routing\Router-&gt;dispatchToRoute()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:625)
    at Illuminate\Routing\Router-&gt;dispatch()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:166)
    at Illuminate\Foundation\Http\Kernel-&gt;Illuminate\Foundation\Http\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/bepsvpt/secure-headers/src/SecureHeadersMiddleware.php:21)
    at Bepsvpt\SecureHeaders\SecureHeadersMiddleware-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21)
    at Illuminate\Foundation\Http\Middleware\TransformsRequest-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php:31)
    at Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21)
    at Illuminate\Foundation\Http\Middleware\TransformsRequest-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php:40)
    at Illuminate\Foundation\Http\Middleware\TrimStrings-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php:27)
    at Illuminate\Foundation\Http\Middleware\ValidatePostSize-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php:86)
    at Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/fideloper/proxy/src/TrustProxies.php:57)
    at Fideloper\Proxy\TrustProxies-&gt;handle()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167)
    at Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103)
    at Illuminate\Pipeline\Pipeline-&gt;then()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:141)
    at Illuminate\Foundation\Http\Kernel-&gt;sendRequestThroughRouter()
(/tmp/Lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:110)
    at Illuminate\Foundation\Http\Kernel-&gt;handle()
(/tmp/Lychee/public/index.php:74)
```

Screenshot 2021-05-15 at 18 36 58

This appears to be because the `findLivePartner` function incorrectly checks the `livePhotoPartner` field on the photo currently being uploaded, which is always `NULL` at this point because we're still in the process of finding the live photo partner. I've updated this to check the local `$livePhotoPartner` var instead. This has fixed the issue for me running Lychee and manually uploading live photos.

Happy to do some more work on the test if needed, but if somebody could point me in the right direction for running the tests that would be much appreciated. I've had a click around at https://lycheeorg.github.io/docs/contributions.html but couldn't see anything about build/test/run.

### CODE REVIEW

**Comment by ildyria** [comment]
First of all, thank you for your contribution. :)

> I've had a click around at lycheeorg.github.io/docs/contributions.html but couldn't see anything about build/test/run.

The CI is automatically run. The thing is since bad actors abused of the free github actions to mine cryptocurrency, we now have to start the CI ourselves for new contributors.

You can run the tests locally by doing `make test` :)
Likewise for the formatting you have `make formatting`.

**Comment by icager** [comment]
Thanks @ildyria for the pointer! I'd completely missed that Makefile, got the tests running locally now so I've added some more assertions to the new test so that it checks the live photo has actually been linked, not just that the upload didn't bug out.

Can't see any changes by `make formatting` so I'm assuming that's all fine.

Please do let me know (no rush though!) if there's anything I should improve/change or if I've missed any of the coding standards. Perfectly happy to make changes.

**Comment by ildyria** [comment]
> if I've missed any of the coding standards.

I haven't had a look at it yet, but `make formatting` ensures that we all have the same coding style (and the CI double check it too). So no worries on that one. :)

**Comment by ildyria** [comment]
@icager I tried to add you as an external contributor to the repo to see if you can start the CI yourself without having us to approve it at each commit. :sweat_smile:

**Comment by icager** [comment]
Thanks @ildyria - looks like I'll need to push a few debug commits as the tests still seem to be failing in CI, so that should really help!

**Comment by ildyria** [comment]
Seems like it is building. :) @icager

**Comment by icager** [comment]
> Seems like it is building. :) @icager

Yep, seems to have done the trick, thanks again for that. Hopefully won't need too many builds to figure it out, looks like it might just be something to do with exiftool versions...

**Comment by ildyria** [comment]
> Hopefully won't need too many builds to figure it out

Take as many builds as you need. :)

**Comment by icager** [comment]
The problem was a renaming of exif fields between the version of `exiftool` I have and the version installed in CI.

I've [raised a PR in the exif lib](https://github.com/LycheeOrg/php-exif/pull/34) to fix the names and hacked the same change in to this PR, which fixes the test. I've taken the debug work back out of this PR, so only that fix to php-exif is pending.

**Comment by ildyria** [comment]
I feel that the composer update broke the RSS part (again).
Thanks Spatie for that...

**Comment by icager** [comment]
Thaks @ildyria for sorting the composer update! This is ready for review as far as I'm concerned - no rush at all, happy to make changes both major and minor.

File: config/feed.php [hunk 1]
```diff
@@ -16,7 +16,7 @@
 			/*
 			 * The feed will be available on this url.
 			 */
-			'url' => '/feed',
+			'url' => '',
```
Comment (nagmat84) [inline config/feed.php:22]: What is the purpose of this change? What are the impacts of changing this from `/feed` to the empty URL?

File: tests/Feature/PhotosTest.php [hunk 2]
```diff
@@ -202,6 +198,56 @@ public function testUpload()
...
+	 * @return void
+	 */
+	public function testLivePhotoUpload()
+	{
+		$photos_tests = new PhotosUnitTest($this);
+
+		AccessControl::log_as_id(0);
+		// MUST use exiftool to get live photo metadata
+		$init_config_value = Configs::get_value('has_exiftool');
+		Configs::set('has_exiftool', '1');
```
Comment (nagmat84) [inline tests/Feature/PhotosTest.php]: What happens if exiftool is not actually available? I guess the test will crash, doesn't it?

File: tests/Feature/PhotosTest.php [hunk 3]
```diff
@@ -277,4 +323,11 @@ public function testImport()
 
 		AccessControl::logout();
 	}
+
+	private function asArray($response)
+	{
+		$content = $response->getContent();
+
+		return json_decode($content);
```
Comment (nagmat84) [inline tests/Feature/PhotosTest.php:336]: The method should not be named `asArray`, because `json_decode` actually returns an object not an array. However, IMHO the method is not necessary at all. See comment above.

File: tests/Feature/PhotosTest.php [hunk 4]
```diff
@@ -202,6 +198,56 @@ public function testUpload()
...
+		);
+
+		$photo_id = $photos_tests->upload($photo_file);
+		$video_id = $photos_tests->upload($video_file);
+
+		$photo = $this->asArray($photos_tests->get($photo_id, 'true'));
+
+		$this->assertEquals($photo_id, $video_id);
+		$this->assertEquals($photo->livePhotoContentID, 'E905E6C6-C747-4805-942F-9904A0281F02');
+		$this->assertStringEndsWith('.mov', $photo->livePhotoUrl);
```
Comment (nagmat84) [inline tests/Feature/PhotosTest.php]: At first, this line puzzled me, because you could not invoke a property `->livePhotoUrl` on an array, if `$photo` was indeed an array as the name of the method `asArray` suggests. However, `asArray` does not return an array so the name of the method is actually not well chosen (see below).

However, the method `asArray` is not required at all. Just take the `TestReponse` which is returned by `$photos_test->get` and use the method `assertJson`. See `GeoDataTest` for examples how to test for particular JSON attrbiutes in a response.

File: tests/Feature/PhotosTest.php [hunk 5]
```diff
@@ -148,14 +148,12 @@ public function testUpload()
 		$photos_tests->dont_see_in_unsorted($id);
 
 		$photos_tests->duplicate($id, 'true');
-		$response = $albums_tests->get($albumID, '', 'true');
-		$content = $response->getContent();
-		$array_content = json_decode($content);
-		$this->assertEquals(2, count($array_content->photos));
+		$album = $this->asArray($albums_tests->get($albumID, '', 'true'));
```
Comment (nagmat84) [inline tests/Feature/PhotosTest.php]: see comment below regarding `asArray`

File: tests/Feature/PhotosTest.php [hunk 6]
```diff
@@ -173,10 +171,8 @@ public function testUpload()
 		// delete the picture after displaying it
 		$photos_tests->delete($ids[1], 'true');
 		$photos_tests->get($id[1], 'false');
-		$response = $albums_tests->get($albumID, '', 'true');
-		$content = $response->getContent();
-		$array_content = json_decode($content);
-		$this->assertEquals(0, count($array_content->photos));
+		$album = $this->asArray($albums_tests->get($albumID, '', 'true'));
```
Comment (nagmat84) [inline tests/Feature/PhotosTest.php]: see comment below regarding `asArray`

**Review by nagmat84 — CHANGES_REQUESTED** [review]
Why has `composer.lock` been changed? Was that accidental?

File: config/feed.php (hunk 1 above)
Comment (ildyria) [reply config/feed.php:22]: This is due to Spacie fucking up... See here: https://github.com/spatie/laravel-feed/issues/151

File: tests/Feature/PhotosTest.php (hunk 2 above)
Comment (ildyria) [reply tests/Feature/PhotosTest.php]: yes. But in our test suite we do install exiftool.

**Comment by ildyria** [comment]
`composer.lock` is changed to apply the update on the php-exif side.

File: tests/Feature/PhotosTest.php (hunk 2 above)
Comment (nagmat84) [reply tests/Feature/PhotosTest.php]: Yes, but the CI at the server side is not the only place where tests are run. Maybe be there are developers which run the test suite locally on their machine, but haven't installed exiftool. Then the tests crash. Therefore I would not change the configuration setting, hut skip the test if exiftool is not available and spit out a warning that not all tests could be run due to missing dependencies.

File: config/feed.php (hunk 1 above)
Comment (nagmat84) [reply config/feed.php:22]: I would advocate to change the comment in the source code above this line and add a few words to the comment telling the reader

 1. that leaving this configuration value empty is intentional,
 2. that the library adds `/feed` by itself and
 3. link to the issue which @ildyria pointed to

File: tests/Feature/PhotosTest.php (hunk 4 above)
Comment (ildyria) [reply tests/Feature/PhotosTest.php]: True, but here we only check two attributes, and one of them is not a direct equality. :)

**Review by nagmat84 — APPROVED** [review]
